src/charm.py

```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/env python3
+import glob
+import subprocess
+import time
+
+from jinja2 import Environment, FileSystemLoader
+import logging
+from pathlib import Path
+
+from kubernetes import client, config
+import kubernetes.client.exceptions
+from oci_image import OCIImageResource
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import ActiveStatus, WaitingStatus, MaintenanceStatus
+
+
+class MetacontrollerOperatorCharm(CharmBase):
+    """Charm the Metacontroller"""
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        if not self.unit.is_leader():
+            self.model.unit.status = WaitingStatus("Waiting for leadership")
+            return
+
+        self.framework.observe(self.on.noop_pebble_ready, self._noop_pebble_ready)
+        self.framework.observe(self.on.install, self._install)
+        self.framework.observe(self.on.remove, self._remove)
+        self.framework.observe(self.on.update_status, self._update_status)
+        # self.framework.observe(self.on.config_changed, self._reconcile)
+
+        self.x = 1
+        self.logger = logging.getLogger(__name__)
+        self.y = 1
+        # TODO: Fix file imports and move ./src/files back to ./files
+        self._manifests_file_root = None
+        self.manifest_file_root = "./src/files/manifests/"
+        self.image = OCIImageResource(self, "oci-image")
+
+    def _noop_pebble_ready(self, _):
+        self.logger.info("noop_pebble_ready fired")
+
+    def _install(self, event):
+        self.logger.info("Installing by instantiating Kubernetes objects")
+        self.unit.status = MaintenanceStatus("Instantiating Kubernetes objects")
+        _, manifests_str = self._render_manifests()
+
+        self.logger.info("Applying manifests")
+
+        completed_process = subprocess.run(
+            ["./kubectl", "apply", "-f-"],
+            input=manifests_str.encode("utf-8"),
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+        )
+        self.logger.info(
+            f"kubectl returned with code '{completed_process.returncode}'"
+            f" and output {completed_process.stdout}"
+        )
+
+        self.logger.info("Waiting for installed Kubernetes objects to be operational")
+        max_attempts = 20
+        for attempt in range(max_attempts):
+            self.logger.info(f"validation attempt {attempt}/{max_attempts}")
+            running = self._check_deployed_resources()
+
+            if running is True:
+                self.logger.info("Resources detected as running")
+                self.logger.info("Install successful")
+                self.unit.status = ActiveStatus()
+                return
+            else:
+                sleeptime = 10
+                self.logger.info(f"Sleeping {sleeptime}s")
+                time.sleep(sleeptime)
```

Comment:
I'm really not sure about this as an approach; note that blocking this hook with a sleep essentially blocks the charm from any future hooks (which I'm sure was intentional), but seems like it'd be better handled with `event.defer` or similar.

Reply:
(Once this is ported to lightkube, this really shouldn't be necessary either, so I'd suggest prioritising that?)

Reply:
I'm not sure about it either and I'm open to ideas, I just don't know a better one.  The intent here was to capture the time between object creation (`kubectl create X`/`lightkube.create`) and everything running (resources created, `status=Ready`, ...).  I don't see how lightkube changes this vs kubectl but I'm curious to hear what you're thinking on that.

I've not used `defer` so I could have this wrong... If we `defer()` here then we (re-)fire the `install` hook next time any other hook is fired, without a time guarantee of when that'll happen.  Worst (practical) case is 5m for the next `update-status`, which seems like a long delay for something that's probably 30s or so wait but not catastrophic.  In that case, would we also need the `install` handler to be "smart" and notice whether this is the first execution (we must create all the things) vs the n-th execution (we should check the things)?  Maybe we it could just attempt to create regardless, handle any duplicated object errors, then always check status after.

Other options:
* emit a custom event (`install-status`) that does the status checking for install.  That event could check once and then defer itself if the objects aren't ready.  This still has the time guarantee problem (deferred events fire whenever something else wants to do work) but it doesn't block the application
* emit a custom event (`install-status`) like above, but if objects aren't ready **it emits a new `install-status` event**.  That will ensure the next status check happens soon and doesn't block the application like sleeping in the handler would, but it means we really spam events.

All options feel kinda bad.  A `defer_for(time='10s')` or `emit_in(time='10s')` might be useful here.

Reply:
So unless I am understanding something wrong, while hook sleeps it does block future hooks, but those hooks remain in a queue and will be executed once our current hook finishes.

In our case, I don't think waiting for the k8s resources to be available is a bad pattern, as either way the other hooks won't do much without those resources being available. What is a bad pattern is sleeping an arbitrary time when we could wait for a resource. With `lightkube` this should be possible with this [PR](https://github.com/gtsystem/lightkube/pull/15).

Reply:
@DomFleischmann I had the same initial reaction but then two worries:
1. If we deploy something slow to install then config change (or even chain a few `juju config A; juju config B`'s together), we'll have a queue of installs that don't all need to happen.  Might be nice to short-circuit the install when it isn't the most recent config anyway
2. afaik sleeping/waiting/whatever does block things like `juju remove-application` calls because of the remove event, so a stalled install cannot be gracefully torn down.  `force` would work (or `--force --no-wait`?  cant remember the difference) but that would be really bad for something that has an important `remove` hook that needs to run (eg: removing `profile-controller` without removing `profiles-crd`).

Is there a way for a charm to ask Juju if there are events pending?  With that we can write our charms to be neighbourly and periodically check if there's something queued behind them to `defer` and get out of the way (or if there's a config change, maybe even kill the current install)

Reply:
@ca-scribner I think a lot of your concerns are just consequences of the event driven pattern that has been chosen for the operator framework.

> If we deploy something slow to install then config change (or even chain a few juju config A; juju config B's together), we'll have a queue of installs that don't all need to happen. Might be nice to short-circuit the install when it isn't the most recent config anyway

If there has been another config change before the first config_changed hook has been executed, the initial hook will apply the latest config. Of course the execution of the second hook is unnecessary but that is not the charms problem, but of the person/thing that changed config that quickly.

> afaik sleeping/waiting/whatever does block things like juju remove-application calls because of the remove event, so a stalled install cannot be gracefully torn down. force would work (or --force --no-wait? cant remember the difference) but that would be really bad for something that has an important remove hook that needs to run (eg: removing profile-controller without removing profiles-crd).

That is true, but at the same time  we are not expecting the charm to sleep forever, there should always be some kind of reasonable timeout, so that the hook ends at some point.

> Is there a way for a charm to ask Juju if there are events pending? With that we can write our charms to be neighbourly and periodically check if there's something queued behind them to defer and get out of the way (or if there's a config change, maybe even kill the current install)

I don't think there is something like that, but what might be possible, is reemiting the event that currently needs to wait and put it at the end of the queue. I'm not entirely sure if that is a pattern we want to pursue, but it might be worth considering, as `defer` most of the times doesn't seem very useful.

Reply:
Re emitting events, I think that comes back to something like emitting a custom `install-status` event that checks/reports on status (or even just emitting an `update-status` event).  That feels like a decent way to get an unblocking, repeated check on status

Reply:
It might be something to consider, for now, I would stick with `lightkube.wait` and a reasonable timeout.

Reply:
This has not been changed in the current PR, but could use `lightkube.wait` once that is released to the wild.  I'm still not sure it actually fixes the underlying problem though.